### PR TITLE
Implement a freeze: unpacker option

### PR DESCRIPTION
**Pull request by casperisfine**
As discussed in https://github.com/msgpack/msgpack-ruby/pull/190

If set to true all parsed objects will be immediately frozen, and strings will be deduplicated if the Ruby implementation allows it.

This feature is meant to be similar to Pysch's `freeze:` option: https://github.com/ruby/psych/pull/414

For now this option is only useful if you plan to keep the parsed objects around, and want to reduce their memory footprint.

But hopefully Ruby 3.0 should soon expose `rb_fstring_new(const char *ptr, long len)` (likely under another name) which should allow to reduce allocations by directly looking up existing strings without allocating a new one.

cc @tagomoris 

NB: I presume that a java implementation is needed, I'm not too well versed into Java by I can try my hand at it. However I'd rather have your feedback on the MRI implementation first.

### CODE REVIEW

**Comment by casperisfine** [comment]
I went ahead and implemented the jruby version.

File: spec/unpacker_spec.rb
Comment (tagomoris) [inline spec/unpacker_spec.rb:667]: Could you add a test for `MessagePack.unpack()` ? We can pass options for Unpackers like `MessagePack.unpack(binary, symbolize_keys: true)`

Comment (casperisfine) [reply spec/unpacker_spec.rb:667]: Sure.

Comment (casperisfine) [reply spec/unpacker_spec.rb:667]: Done.

File: ext/java/org/msgpack/jruby/Decoder.java
Comment (tagomoris) [inline ext/java/org/msgpack/jruby/Decoder.java:125]: `Ruby#freezeAndDedupString` looks to freeze&dedup the string, not only dedup.
Is there any reason to call `setFrozen(true)` explicitly here?

**Review by tagomoris — APPROVED** [review]
Most things look good. Great!
I left a minor comment about how to get frozen strings in JRuby. Could you check that?

File: ext/java/org/msgpack/jruby/Decoder.java
Comment (casperisfine) [reply ext/java/org/msgpack/jruby/Decoder.java:125]: Yes, today it's a bit useless, however in theory if you call it with a frozen string, it could directly intern that string without duplicating it first. That's what MRI 3.0 now does, see: https://github.com/ruby/spec/pull/788

I do hope to change that behavior in JRuby at some point, so I figured we might as well call it with a frozen string directly.

File: ext/msgpack/unpacker.c
Comment (casperisfine) [inline ext/msgpack/unpacker.c:186]: Please don't merge just yet, this part is not quite correct. I'll push a fix in a couple minutes.

Comment (casperisfine) [reply ext/msgpack/unpacker.c:187]: It's fixed now.

**Comment by tagomoris** [comment]
Plz ping me when you solve the problem (and failing CI).

**Comment by casperisfine** [comment]
> Plz ping me when you solve the problem (and failing CI).

It should be good now. CI is green. Sorry for the last change.

File: ext/msgpack/unpacker.c
Comment (tagomoris) [reply ext/msgpack/unpacker.c:187]: I can't see the commit to fix this (but the whole commit forcibly pushed).
Could you explain what was wrong and fixed? I want to understand it.

Comment (casperisfine) [reply ext/msgpack/unpacker.c:187]: Of course. So the `String#-@` behavior changed a lot on every single version.

Up to 2.5, it had a simple guard statement `return self if frozen?`. Note that frozen doesn't mean "interned". So if you want to deduplicate a string in 2.5 or older, you have to supply a mutable string.

2.6 and 2.7 are a bit smarter, and will deduplicate a supplied string even if it is frozen, however the code is a bit like: `str = str.dup if str.frozen?`, meaning that calling `-@` on a frozen string doesn't give you anything compared to a mutable one.

Then starting in 2.8 (named 3.0 now), if the string is frozen it can directly be put into the interned string table, so freezing the string before calling `-@` will save you an allocation per unique string.

Now my feature checking condition I pushed was a bit broken. The intent was to test wether frozen string could be directly promoted to interned string.

Here's the first version of my test:

```ruby
  a = -(%w(t e s t).join.freeze)
  b = -(%w(t e s t).join.freeze)
  if a.equal?(b)
```

This check was buggy because on 2.6 and 2.7, both expression would return a third instance that was previously created an interned. No huge deal, but it mean we'd do one useless `rb_str_freeze` call for each string.

So I rewrote it as:

```ruby
  s = rand.to_s.freeze
  if (-s).equal?(s)
```

Which did properly work on 2.6+, however it was broken on 2.5- because while the returned string was the same instance, it wasn't interned at all (because `return self if frozen?`).

So I finally pushed:

```ruby
  s = rand.to_s.freeze
  if (-s).equal?(s) && (-s.dup).equal?(s)
```

Which properly check that frozen strings can be promoted to interned strings.

I hope I explained this properly.

Comment (tagomoris) [reply ext/msgpack/unpacker.c:187]: That makes sense, perfectly. Thanks!

**Comment by tagomoris** [comment]
I'll merge this change tomorrow if there's no further commits/comments.
The release will be issued after merging #193

**Comment by casperisfine** [comment]
Works for me, thanks!

**Pull request merged**
